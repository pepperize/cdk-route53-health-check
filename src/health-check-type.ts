/**
 * The type of health check to create.
 *
 * @internal
 */
export enum HealthCheckType {
  HTTP = "HTTP",
  HTTPS = "HTTPS",
  HTTP_STR_MATCH = "HTTP_STR_MATCH",
  HTTPS_STR_MATCH = "HTTPS_STR_MATCH",
  TCP = "TCP",
  CALCULATED = "CALCULATED",
  CLOUDWATCH_METRIC = "CLOUDWATCH_METRIC",
  RECOVERY_CONTROL = "RECOVERY_CONTROL",
}
