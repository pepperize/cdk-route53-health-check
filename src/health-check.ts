import { ITaggable, Resource, TagManager, TagType } from "aws-cdk-lib";
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch";
import * as route53 from "aws-cdk-lib/aws-route53";
import { RecordSet } from "aws-cdk-lib/aws-route53";
import { Construct } from "constructs";

/**
 * @internal
 */
export interface HealthCheckOptions {
  /**
   * The display name of this Route53 HealthCheck
   */
  readonly healthCheckName?: string;
  /**
   * Whether to invert the status of the Route53 health check status.
   */
  readonly inverted?: boolean;
}

export interface IHealthCheck {
  readonly healthCheckId: string;
  /**
   * Sets `this.healthCheckId` as the value for `HealthCheckId` on the given RecordSet.
   *
   * <b>Applies only to alias, failover alias, geolocation alias, latency alias, and weighted alias resource record sets</b>
   *
   * @param recordSet The Route53 RecordSet to configure failover
   * @param evaluateTargetHealth Inherit the health of the referenced Alias RecordSet Target
   * @param failover Sets `PRIMARY` or `SECONDARY` as the value for `Failover` on the given RecordSet.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-route53-aliastarget.html#cfn-route53-aliastarget-evaluatetargethealth
   */
  failover(recordSet: RecordSet, evaluateTargetHealth?: boolean, failover?: Failover): void;
  /**
   * Sets `PRIMARY` as the value for `Failover` on the given RecordSet. Additionally, sets `this.healthCheckId` as the value for `HealthCheckId`.
   *
   * <b>Applies only to alias, failover alias, geolocation alias, latency alias, and weighted alias resource record sets</b>
   *
   * @param recordSet The Route53 RecordSet to configure failover
   * @param evaluateTargetHealth Inherit the health of the referenced Alias RecordSet Target
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-route53-aliastarget.html#cfn-route53-aliastarget-evaluatetargethealth
   */
  failoverPrimary(recordSet: route53.RecordSet, evaluateTargetHealth?: boolean): void;
  /**
   * Sets `PRIMARY` as the value for `Failover` on the given RecordSet. Additionally, sets `this.healthCheckId` as the value for `HealthCheckId`.
   *
   * <b>Applies only to alias, failover alias, geolocation alias, latency alias, and weighted alias resource record sets</b>
   *
   * @param recordSet The Route53 RecordSet to configure failover
   * @param evaluateTargetHealth Inherit the health of the referenced Alias RecordSet Target
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-route53-aliastarget.html#cfn-route53-aliastarget-evaluatetargethealth
   */
  failoverSecondary(recordSet: route53.RecordSet, evaluateTargetHealth?: boolean): void;
  /**
   * Return the given named metric for this HealthCheck.
   */
  metric(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric;

  /**
   * Route53 health checkers report that the HealthCheck is healthy or unhealthy.
   */
  metricHealthCheckStatus(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
}

/**
 * @internal
 */
export abstract class HealthCheckBase extends Resource implements IHealthCheck, ITaggable {
  /**
   * Import an existing Route53 HealthCheck.
   */
  public static fromHealthCheckId(scope: Construct, id: string, healthCheckId: string): IHealthCheck {
    return new (class Imported extends HealthCheckBase {
      public readonly healthCheckId = healthCheckId;
      constructor() {
        super(scope, id);
      }
    })();
  }

  public abstract readonly healthCheckId: string;

  readonly tags = new TagManager(TagType.STANDARD, "AWS::Route53::HealthCheck");

  failover(recordSet: RecordSet, evaluateTargetHealth?: boolean, failover?: Failover) {
    const resource = recordSet.node.defaultChild;

    if (!(resource instanceof route53.CfnRecordSet)) {
      throw new Error(`RecordSet ${recordSet.node.path} doesn't have a default child of type CfnRecordSet`);
    }

    resource.healthCheckId = this.healthCheckId;

    if (failover) {
      resource.failover = failover;
    }

    if (evaluateTargetHealth) {
      resource.addPropertyOverride("AliasTarget.EvaluateTargetHealth", evaluateTargetHealth);
    }
  }

  failoverPrimary(recordSet: route53.RecordSet, evaluateTargetHealth?: boolean) {
    this.failover(recordSet, evaluateTargetHealth, Failover.PRIMARY);
  }

  failoverSecondary(recordSet: route53.RecordSet, evaluateTargetHealth?: boolean) {
    this.failover(recordSet, evaluateTargetHealth, Failover.SECONDARY);
  }

  /**
   * Return the given named metric for this HealthCheck.
   */
  metric(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return new cloudwatch.Metric({
      namespace: "AWS/Route53",
      metricName,
      dimensionsMap: { HealthCheckId: this.healthCheckId },
      ...props,
    }).attachTo(this);
  }
  /**
   * Route53 health checkers report that the HealthCheck is healthy or unhealthy.
   */
  metricHealthCheckStatus(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.metric("HealthCheckStatus", { statistic: cloudwatch.Statistic.MINIMUM, ...props });
  }
}

export enum Failover {
  /**
   * The primary record set
   */
  PRIMARY = "PRIMARY",
  /**
   * The secondary record set
   */
  SECONDARY = "SECONDARY",
}
