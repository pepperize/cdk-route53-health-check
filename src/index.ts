export * from "./alarm-health-check";
export * from "./calculated-health-check";
export * from "./endpoint-health-check";
export * from "./health-check";
