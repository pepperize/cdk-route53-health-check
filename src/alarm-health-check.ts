import { ArnFormat, Stack, Tags } from "aws-cdk-lib";
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch";
import * as route53 from "aws-cdk-lib/aws-route53";
import { Construct } from "constructs";
import { HealthCheckBase, HealthCheckOptions } from "./health-check";
import { HealthCheckType } from "./health-check-type";

export interface AlarmHealthCheckProps extends HealthCheckOptions {
  /**
   * The alarm that Route53 monitors
   */
  readonly alarm: cloudwatch.IAlarm;
  /**
   * The status to assign to the HealthCheck when CloudWatch has insufficient data about the metric
   */
  readonly insufficientDataHealthStatus?: InsufficientDataHealthStatus;
}

/**
 * Create a Route53 HealthCheck that monitors a CloudWatch Alarm.
 *
 * <b>Example</b>
 * ```typescript
 * const alarm new Alarm(stack, "Alarm", {
 *   // ...
 * });
 * new AlarmHealthCheck(stack, "HealthCheck", {
 *   alarm: alarm,
 * });
 * ```
 * @link https://docs.aws.amazon.com/de_de/AWSCloudFormation/latest/UserGuide/aws-resource-route53-healthcheck.html#aws-resource-route53-healthcheck-properties
 *
 * @resource AWS::Route53::HealthCheck
 */
export class AlarmHealthCheck extends HealthCheckBase {
  public readonly healthCheckId: string;

  constructor(scope: Construct, id: string, props: AlarmHealthCheckProps) {
    super(scope, id);

    const { alarmName, alarmArn } = props.alarm;
    // https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazoncloudwatch.html
    const alarmRegion = Stack.of(props.alarm).splitArn(alarmArn, ArnFormat.COLON_RESOURCE_NAME).region;

    const insufficientDataHealthStatus =
      props.insufficientDataHealthStatus ?? InsufficientDataHealthStatus.LAST_KNOWN_STATUS;

    const resource = new route53.CfnHealthCheck(this, "Resource", {
      healthCheckConfig: {
        alarmIdentifier: {
          name: alarmName,
          region: alarmRegion!,
        },
        insufficientDataHealthStatus: insufficientDataHealthStatus,
        type: HealthCheckType.CLOUDWATCH_METRIC,
      },
      healthCheckTags: this.tags.renderedTags,
    });

    this.healthCheckId = resource.attrHealthCheckId;

    if (props.healthCheckName) {
      Tags.of(this).add("Name", props.healthCheckName);
    }
  }
}

export enum InsufficientDataHealthStatus {
  /**
   * Route53 considers the health check to be healthy.
   */
  HEALTHY = "Healthy",
  /**
   * Route53 considers the health check to be unhealthy.
   */
  UNHEALTHY = "Unhealthy",
  /**
   * Route53 uses the status of the health check from the last time that CloudWatch had sufficient data to determine the alarm state, otherwise healthy
   */
  LAST_KNOWN_STATUS = "LastKnownStatus",
}
