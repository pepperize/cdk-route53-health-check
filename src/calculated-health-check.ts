import { Annotations, Lazy, Tags } from "aws-cdk-lib";
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch";
import * as route53 from "aws-cdk-lib/aws-route53";
import { Construct } from "constructs";
import { HealthCheckBase, HealthCheckOptions, HealthCheckType, IHealthCheck } from "./health-check";

export interface CalculatedHealthCheckProps extends HealthCheckOptions {
  /**
   * The list of HealthCheck that monitors other Route53 HealthChecks.
   */
  readonly childHealthChecks?: IHealthCheck[];
  /**
   * The number of child HealthChecks that Amazon Route53 must consider healthy
   */
  readonly healthThreshold?: number;
}

/**
 * Create a Route53 HealthCheck that monitors other Route53 HealthChecks.
 *
 * <b>Example</b>
 * ```typescript
 * const healthCheck = new EndpointHealthCheck(stack, "HealthCheck", {
 *   domainName: "pepperize.com",
 * });
 * new CalculatedHealthCheck(stack, "CalculatedHealthCheck", {
 *   childHealthChecks: [healthCheck],
 * });
 * ```
 * @link https://docs.aws.amazon.com/de_de/AWSCloudFormation/latest/UserGuide/aws-resource-route53-healthcheck.html#aws-resource-route53-healthcheck-properties
 */
export class CalculatedHealthCheck extends HealthCheckBase {
  public readonly healthCheckId: string;

  private childHealthChecks: IHealthCheck[];

  constructor(scope: Construct, id: string, props: CalculatedHealthCheckProps) {
    super(scope, id);

    this.childHealthChecks = props.childHealthChecks ?? [];

    if (this.childHealthChecks.length > 256) {
      Annotations.of(this).addError("ChildHealthChecks has to be smaller than 256");
    }

    if (undefined != props.healthThreshold && props.healthThreshold > 256) {
      Annotations.of(this).addError("HealthThreshold has to be smaller than 256");
    }

    const resource = new route53.CfnHealthCheck(this, "Resource", {
      healthCheckConfig: {
        childHealthChecks: Lazy.list({
          produce: () => this.childHealthChecks.map((healthCheck) => healthCheck.healthCheckId),
        }),
        healthThreshold: props.healthThreshold,
        type: HealthCheckType.CALCULATED,
      },
      healthCheckTags: this.tags.renderedTags,
    });

    this.healthCheckId = resource.attrHealthCheckId;

    if (props.healthCheckName) {
      Tags.of(this).add("Name", props.healthCheckName);
    }
  }

  /**
   * The number of ChildHealthChecks that are healthy that Route53 is monitoring.
   *
   * Valid statistics: Average (recommended), Minimum, Maximum
   */
  public metricChildHealthCheckHealthyCount(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.metric("ChildHealthCheckHealthyCount", { statistic: cloudwatch.Statistic.AVERAGE, ...props });
  }

  public addChildHealthCheck(healthCheck: IHealthCheck) {
    this.childHealthChecks.push(healthCheck);
  }
}
