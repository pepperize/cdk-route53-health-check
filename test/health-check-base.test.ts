import { Stack } from "aws-cdk-lib";
import { Match, Template } from "aws-cdk-lib/assertions";
import * as cloudfront from "aws-cdk-lib/aws-cloudfront";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as elasticloadbalancingv2 from "aws-cdk-lib/aws-elasticloadbalancingv2";
import * as route53 from "aws-cdk-lib/aws-route53";
import * as targets from "aws-cdk-lib/aws-route53-targets";
import { EndpointHealthCheck } from "../src";

describe("HealthCheckBase", () => {
  it("Should configure DNS failover", () => {
    // Given
    const stack = new Stack();

    const hostedZone = new route53.PublicHostedZone(stack, "HostedZone", {
      zoneName: "pepperize.com",
    });

    const distribution = cloudfront.Distribution.fromDistributionAttributes(stack, "Distribution", {
      domainName: "pepperize.com",
      distributionId: "d111111abcdef8",
    });
    const recordSetPrimary = new route53.ARecord(stack, "RecordSetPrimary", {
      recordName: "www.pepperize.com",
      zone: hostedZone,
      target: route53.RecordTarget.fromAlias(new targets.CloudFrontTarget(distribution)),
    });

    const healthCheckPrimary = new EndpointHealthCheck(stack, "HealthCheckPrimary", {
      domainName: "www.pepperize.com",
    });
    healthCheckPrimary.failoverPrimary(recordSetPrimary);

    const alb = new elasticloadbalancingv2.ApplicationLoadBalancer(stack, "ApplicationLoadBalancer", {
      vpc: new ec2.Vpc(stack, "Vpc"),
    });
    const recordSetSecondary = new route53.ARecord(stack, "RecordSetSecondary", {
      recordName: "www-1.pepperize.com",
      zone: hostedZone,
      target: route53.RecordTarget.fromAlias(new targets.LoadBalancerTarget(alb)),
    });
    const healthCheckSecondary = new EndpointHealthCheck(stack, "HealthCheckSecondary", {
      domainName: "www-1.pepperize.com",
    });
    healthCheckSecondary.failoverSecondary(recordSetSecondary, true);

    // When
    const template = Template.fromStack(stack);

    // Then
    expect(template).toMatchSnapshot();

    template.hasResourceProperties("AWS::Route53::RecordSet", {
      AliasTarget: Match.not({
        EvaluateTargetHealth: true,
      }),
      Failover: "PRIMARY",
      HealthCheckId: {
        "Fn::GetAtt": Match.arrayWith([Match.stringLikeRegexp("HealthCheckPrimary"), "HealthCheckId"]),
      },
      Name: "www.pepperize.com.",
      Type: "A",
    });

    template.hasResourceProperties("AWS::Route53::RecordSet", {
      AliasTarget: {
        EvaluateTargetHealth: true,
      },
      Failover: "SECONDARY",
      HealthCheckId: {
        "Fn::GetAtt": Match.arrayWith([Match.stringLikeRegexp("HealthCheckSecondary"), "HealthCheckId"]),
      },
      Name: "www-1.pepperize.com.",
      Type: "A",
    });
  });
});
