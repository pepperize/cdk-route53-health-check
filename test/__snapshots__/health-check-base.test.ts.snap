// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HealthCheckBase Should configure DNS failover 1`] = `
Object {
  "Mappings": Object {
    "AWSCloudFrontPartitionHostedZoneIdMap": Object {
      "aws": Object {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": Object {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "ApplicationLoadBalancerFD56DEE1": Object {
      "Properties": Object {
        "LoadBalancerAttributes": Array [
          Object {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internal",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "ApplicationLoadBalancerSecurityGroup4BBE5415",
              "GroupId",
            ],
          },
        ],
        "Subnets": Array [
          Object {
            "Ref": "VpcPrivateSubnet1Subnet536B997A",
          },
          Object {
            "Ref": "VpcPrivateSubnet2Subnet3788AAA1",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "ApplicationLoadBalancerSecurityGroup4BBE5415": Object {
      "Properties": Object {
        "GroupDescription": "Automatically created Security Group for ELB ApplicationLoadBalancer",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "HealthCheckPrimaryCE09F256": Object {
      "Properties": Object {
        "HealthCheckConfig": Object {
          "EnableSNI": true,
          "FullyQualifiedDomainName": "www.pepperize.com",
          "Port": 443,
          "Type": "HTTPS",
        },
        "HealthCheckTags": Array [
          Object {
            "Key": "Name",
            "Value": "www.pepperize.com",
          },
        ],
      },
      "Type": "AWS::Route53::HealthCheck",
    },
    "HealthCheckSecondary019437FE": Object {
      "Properties": Object {
        "HealthCheckConfig": Object {
          "EnableSNI": true,
          "FullyQualifiedDomainName": "www-1.pepperize.com",
          "Port": 443,
          "Type": "HTTPS",
        },
        "HealthCheckTags": Array [
          Object {
            "Key": "Name",
            "Value": "www-1.pepperize.com",
          },
        ],
      },
      "Type": "AWS::Route53::HealthCheck",
    },
    "HostedZoneDB99F866": Object {
      "Properties": Object {
        "Name": "pepperize.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "RecordSetPrimaryA4D70047": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": "pepperize.com",
          "HostedZoneId": Object {
            "Fn::FindInMap": Array [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              Object {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "Failover": "PRIMARY",
        "HealthCheckId": Object {
          "Fn::GetAtt": Array [
            "HealthCheckPrimaryCE09F256",
            "HealthCheckId",
          ],
        },
        "HostedZoneId": Object {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "www.pepperize.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "RecordSetSecondary6B17D1A8": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::Join": Array [
              "",
              Array [
                "dualstack.",
                Object {
                  "Fn::GetAtt": Array [
                    "ApplicationLoadBalancerFD56DEE1",
                    "DNSName",
                  ],
                },
              ],
            ],
          },
          "EvaluateTargetHealth": true,
          "HostedZoneId": Object {
            "Fn::GetAtt": Array [
              "ApplicationLoadBalancerFD56DEE1",
              "CanonicalHostedZoneID",
            ],
          },
        },
        "Failover": "SECONDARY",
        "HealthCheckId": Object {
          "Fn::GetAtt": Array [
            "HealthCheckSecondary019437FE",
            "HealthCheckId",
          ],
        },
        "HostedZoneId": Object {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "www-1.pepperize.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "Vpc8378EB38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VpcIGWD7BA715C": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VpcPrivateSubnet1DefaultRouteBE02A9ED": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "VpcPublicSubnet1NATGateway4D7517AA",
        },
        "RouteTableId": Object {
          "Ref": "VpcPrivateSubnet1RouteTableB2C5B500",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPrivateSubnet1RouteTableAssociation70C59FA6": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPrivateSubnet1RouteTableB2C5B500",
        },
        "SubnetId": Object {
          "Ref": "VpcPrivateSubnet1Subnet536B997A",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPrivateSubnet1RouteTableB2C5B500": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPrivateSubnet1Subnet536B997A": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcPrivateSubnet2DefaultRoute060D2087": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "VpcPublicSubnet2NATGateway9182C01D",
        },
        "RouteTableId": Object {
          "Ref": "VpcPrivateSubnet2RouteTableA678073B",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPrivateSubnet2RouteTableA678073B": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPrivateSubnet2RouteTableAssociationA89CAD56": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPrivateSubnet2RouteTableA678073B",
        },
        "SubnetId": Object {
          "Ref": "VpcPrivateSubnet2Subnet3788AAA1",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPrivateSubnet2Subnet3788AAA1": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcPublicSubnet1DefaultRoute3DA9E72A": Object {
      "DependsOn": Array [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet1RouteTable6C95E38E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPublicSubnet1EIPD7E02669": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "VpcPublicSubnet1NATGateway4D7517AA": Object {
      "DependsOn": Array [
        "VpcPublicSubnet1DefaultRoute3DA9E72A",
        "VpcPublicSubnet1RouteTableAssociation97140677",
      ],
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "VpcPublicSubnet1EIPD7E02669",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet1Subnet5C2D37C4",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "VpcPublicSubnet1RouteTable6C95E38E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPublicSubnet1RouteTableAssociation97140677": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet1RouteTable6C95E38E",
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet1Subnet5C2D37C4",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPublicSubnet1Subnet5C2D37C4": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcPublicSubnet2DefaultRoute97F91067": Object {
      "DependsOn": Array [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet2RouteTable94F7E489",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPublicSubnet2EIP3C605A87": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "VpcPublicSubnet2NATGateway9182C01D": Object {
      "DependsOn": Array [
        "VpcPublicSubnet2DefaultRoute97F91067",
        "VpcPublicSubnet2RouteTableAssociationDD5762D8",
      ],
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "VpcPublicSubnet2EIP3C605A87",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet2Subnet691E08A3",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "VpcPublicSubnet2RouteTable94F7E489": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPublicSubnet2RouteTableAssociationDD5762D8": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet2RouteTable94F7E489",
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet2Subnet691E08A3",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPublicSubnet2Subnet691E08A3": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "Default/Vpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcVPCGWBF912B6E": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
